src/noteburst/jupyterclient/jupyterlab.py

```diff
@@ -0,0 +1,708 @@
+                    progress=event_dict["progress"],
+                    message=event_dict["message"],
+                    ready=event_dict.get("ready", False),
+                )
+            except Exception as e:
+                msg = f"Ignoring invalid progress event: {raw_event}: {str(e)}"
+                self._logger.warning(
+                    msg, raw_event=raw_event, exception=str(e)
+                )
+                continue
+
+            # Log and yield the event
+            now = datetime.datetime.now(tz=datetime.timezone.utc)
+            elapsed = int((now - self._start).total_seconds())
+            if event.ready:
+                status = "complete"
+            else:
+                status = "in progress"
+            self._logger.info(
+                "Spawning",
+                status=status,
+                elapsed=elapsed,
+                details=event.message,
+            )
+            yield event
+
+
+class JupyterLabSessionStreamComplete(Exception):
+    """Raised when the full result has been recieved from a JupyterLab
+    session websocket iterator.
+    """
+
+
+@dataclass(frozen=True)
+class JupyterLabSession:
+    """Represents an open session with a JupyterLab.
+
+    This holds the information a client needs to talk to JupyterLab in order to
+    execute code.
+    """
+
+    username: str
+    session_id: str
+    kernel_id: str
+    websocket: WebSocketClientProtocol
+    logger: BoundLogger
+
+    async def run_python(self, code: str) -> str:
+        """Run a Python code in the JupyterLab kernel."""
+        msg_id = uuid4().hex
+        msg = {
+            "header": {
+                "username": self.username,
+                "version": "5.0",
+                "session": self.session_id,
+                "msg_id": msg_id,
+                "msg_type": "execute_request",
+            },
+            "parent_header": {},
+            "channel": "shell",
+            "content": {
+                "code": code,
+                "silent": False,
+                "store_history": False,
+                "user_expressions": {},
+                "allow_stdin": False,
+            },
+            "metadata": {},
+            "buffers": {},
+        }
+
+        await self.websocket.send(json.dumps(msg))
+
+        result = ""
+        async for message in self.websocket:
+            try:
+                result += await self._process_run_python_message(
+                    message, msg_id, code
+                )
+            except CodeExecutionError as e:
+                if result:
+                    error_message = f"{result}\nError:\n{str(e.error)}"
+                else:
+                    error_message = str(e)
+                raise CodeExecutionError(
+                    username=e.username,
+                    code=e.code,
+                    code_type=e.code_type,
+                    error=error_message,
+                    status=e.status,
+                )
```

Comment:
I like that you broke this function up into two functions, but the cost of additional code to do this annotation of the exception is kind of unfortunate. To make this read more clearly, I would add a result attribute to `CodeExecutionError` that lets you add the result output separately, use it in the `__str__` override for that exception, and then use something like:
```
except CodeExecutionError as e:
    if result:
        e.result = result
    raise
```